app/pages/Recovery/util.ts

```diff
@@ -0,0 +1,304 @@
+ * @param global: current global parameters.
+ * @param startingCredNumber: credentialNumber, from which to start attempting to recover credentials from.
+ * @returns Returns an object containing the list of all recovered credentials and their accounts. The length of these lists are always the same, and each account matches the credential on the same index.
+ */
+export async function recoverCredentials(
+    prfKeySeed: string,
+    identityId: number,
+    blockHash: string,
+    global: Global,
+    startingCredNumber = 0
+) {
+    const credentials = [];
+    const accounts = [];
+    let credNumber = startingCredNumber;
+    while (credNumber < maxCredentialsOnAccount) {
+        const credId = await getCredId(prfKeySeed, credNumber, global);
+
+        const recovered = await recoverCredential(
+            credId,
+            blockHash,
+            credNumber,
+            identityId
+        );
+
+        if (recovered) {
+            credentials.push(recovered.credential);
+            accounts.push(recovered.account);
+        }
+        credNumber += 1;
+    }
+
+    return { credentials, accounts };
+}
+
+/**
+ * Imports a list of accounts, but only non-duplicates.
+ * @param accounts: the accounts to be added.
+ * @param addressBook: the addressBook is used to check for duplicates.
+ */
+export async function addAccounts(
+    accounts: Account[],
+    addressBook: AddressBookEntry[]
+) {
+    for (const account of accounts) {
+        const { address } = account;
+        const accountExists = (await findAccounts({ address })).length > 0;
+        if (!accountExists) {
+            importAccount(account);
+            if (!addressBook.some((abe) => abe.address === address)) {
+                importAddressBookEntry({
+                    readOnly: true,
+                    name: account.name,
+                    address,
+                    note: 'Recovered account',
+                });
+            }
+        }
+    }
+}
+
+/**
+ * Attempts to recover credentials on an identity.
+ * @param prfKeySeed: Seed of the prfKey of the identity.
+ * @param blockHash: block at which the function recover credentials.
+ * @param global: current global parameters.
+ * @param identityId: id of the identity.
+ * @param addressBook: the addressBook is used to check for duplicates when inserting new accounts.
+ * @returns Returns the recovered accounts.
+ */
+export async function recoverFromIdentity(
+    prfKeySeed: string,
+    blockHash: string,
+    global: Global,
+    identityId: number,
+    addressBook: AddressBookEntry[]
+) {
+    const nextCredentialNumber = await getNextCredentialNumber(identityId);
+    const { credentials, accounts } = await recoverCredentials(
+        prfKeySeed,
+        identityId,
+        blockHash,
+        global,
+        nextCredentialNumber
+    );
+
+    if (accounts.length > 0) {
+        await addAccounts(accounts, addressBook);
```

Comment:
Adding the accounts and credentials should happen transactionally, at least for each of the corresponding pairs.

Reply:
I have done this